plugins/languages/treesitter/treesitter.nix

```
@@ -6,210 +6,264 @@
+      '';
+    };
+
+    nixGrammars = mkOption {
+      type = types.bool;
+      default = true;
+      description = "Whether to automatically install grammar packages.";
+    };
+
+    nixvimInjections = mkOption {
+      type = types.bool;
+      default = true;
+      description = ''Whether to enable Nixvim injections, such as recognising extraConfigLua as lua.'';
+    };
+  };
+
+  callSetup = false;
+  extraConfig = cfg: {
+    assertions = [
+      {
+        assertion = cfg.nixGrammars -> cfg.package != null;
+        message = ''
+          Nixvim (plugins.treesitter): You have defined nixGrammars, but `plugins.treesitter.package` is `null`.
+          This package is required to include grammars to install.
         '';
-      };
+      }
+      {
+        assertion =
+          ((builtins.length cfg.grammarPackages > 0) || cfg.grammarPackages == "all") -> cfg.package != null;
+        message = ''
+          Nixvim (plugins.treesitter): You have defined grammarPackages, but `plugins.treesitter.package` is `null`
+          This package is required to choose which grammars to install.
+        '';
+      }
+      {
+        assertion = !cfg.nixGrammars -> cfg.gccPackage != null;
+        message = ''
+          Nixvim (plugins.treesitter): You have disabled nixGrammars, but `plugins.treesitter.gccPackage` is `null`
+          This package is required to build the grammars at runtime.
+        '';
+      }
+    ];
 
-      extraPlugins =
-        if cfg.nixGrammars then [ (cfg.package.withPlugins (_: cfg.grammarPackages)) ] else [ cfg.package ];
-      extraPackages = with pkgs; [
-        tree-sitter
-        nodejs
-        cfg.gccPackage
-      ];
+    extraConfigLua =
+      # NOTE: It MUST be at the beginning of runtimepath. Otherwise the parsers from Neovim itself
+      # is loaded that may not be compatible with the queries from the 'nvim-treesitter' plugin.
+      (optionalString (cfg.settings.parser_install_dir != null) ''
+        vim.opt.runtimepath:prepend("${cfg.settings.parser_install_dir}")
+      '')
+      + ''
+        require('nvim-treesitter.configs').setup(${helpers.toLuaObject cfg.settings})
+      ''
+      + (optionalString (cfg.languageRegister != { }) ''
+        __parserFiletypeMappings = ${helpers.toLuaObject cfg.languageRegister}
 
-      opts = mkIf cfg.folding {
-        foldmethod = "expr";
-        foldexpr = "nvim_treesitter#foldexpr()";
-      };
+        for parser_name, ft in pairs(__parserFiletypeMappings) do
+          require('vim.treesitter.language').register(parser_name, ft)
+        end
+      '');
+
+    extraFiles = mkIf cfg.nixvimInjections {
+      "queries/nix/injections.scm" = readFile ./injections.scm;
     };
+
+    extraPlugins = mkIf (cfg.package != null) (mkMerge [
+      (mkIf cfg.nixGrammars [ (cfg.package.withPlugins (_: cfg.grammarPackages)) ])
+      (mkIf (!cfg.nixGrammars) [ cfg.package ])
+    ]);
```

Because it's a list, we can simplify this and avoid using mkMerge:
```suggestion
    extraPlugins = mkIf (cfg.package != null) [
      (mkIf cfg.nixGrammars (cfg.package.withPlugins (_: cfg.grammarPackages)))
      (mkIf (!cfg.nixGrammars) cfg.package)
    ];
```